broadway_api/utils/bootstrap.py

```diff
@@ -3,44 +3,85 @@
 import logging
 import os
 import sys
-import uuid
+import signal
+
+from typing import Dict, Any
+from queue import Queue
 
 from pymongo import MongoClient
 from pymongo.errors import ConnectionFailure
 
+import tornado
+import tornado.web
+import tornado.ioloop
+import tornado.httpserver
+
+from logging.handlers import TimedRotatingFileHandler
+
 from broadway_api.definitions import course_config
 from broadway_api.daos import CourseDao, WorkerNodeDao
 from broadway_api.models import Course
 
-import config
+import broadway_api.callbacks as callbacks
+import broadway_api.handlers.client as client_handlers
+import broadway_api.handlers.worker as worker_handlers
+import broadway_api.handlers.worker_ws as worker_ws_handlers
+
+logger = logging.getLogger("broadway.bootstrap")
+
+
+def initialize_global_settings(flags: Dict[str, Any]) -> Dict[str, Any]:
+    return {
+        "CLUSTER_TOKEN": flags["token"],
+        "FLAGS": flags,
+        "DB": None,
+        "QUEUE": Queue(),
+        "WS_CONN_MAP": {},
+    }
+
+
+def initialize_logger(settings: Dict[str, Any], flags: Dict[str, Any]):
+    log_dir = flags["log_dir"]
+    log_level = flags["log_level"]
+    log_rotate = flags["log_rotate"]
+    log_backup = flags["log_backup"]
+
+    os.makedirs(log_dir, exist_ok=True)
 
-logger = logging.getLogger("bootstrap")
+    rotating_handler = TimedRotatingFileHandler(
+        "{}/log".format(log_dir), when=log_rotate, backupCount=log_backup
+    )
 
+    logging.basicConfig(
+        handlers=[rotating_handler, logging.StreamHandler()],
+        format="%(asctime)s %(levelname)s %(module)s.%(funcName)s: %(message)s",
+        level=log_level,
+    )
 
-def initialize_cluster_token():
-    if os.environ.get("BROADWAY_TOKEN"):
-        logger.info("using cluster token from environment")
-        token = os.environ.get("BROADWAY_TOKEN")
-    else:
-        logger.info("generating cluster token")
-        token = str(uuid.uuid4())
-        logger.info("cluster token is {}".format(token))
-    return token
+    logging.getLogger("tornado.access").addHandler(rotating_handler)
+    logging.getLogger("tornado.access").propagate = False
 
+    logging.getLogger("tornado.general").addHandler(rotating_handler)
```

Comment:
I am slightly confused which the rotating handler needs to be added twice as `tornado.general` and `tornado.access`. Would that not double the number of logs generated from the rotating handler.?

Reply:
umm based on my limited understanding of python logging: I think those loggers' handler list defaults to [], so adding a handler would not produce duplicated logs. and since I'm preventing the logs from propagating to the root handler, if this is not set, lastResort will be used which defaults to printing to stderr

Reply:
oops just realized just using `tornado` would also work..

Reply:
Oh so could we just use `tornado` then?

Reply:
Looks like you fixed it, resolving this.